File: rigidbodytransform/rigidtransform_test.py
import time
import numpy as np
from rigidtransform import \
    rotm2quat, quat2rotm, axang2quat, \
    quat2axang, axang2rotm, rotm2axang, \
    rvec2axang, rvec2quat, rvec2rotm, \
    axang2rvec, quat2rvec, rotm2rvec
from quatmath import quatrotate

import cv2


if __name__ == '__main__':

    A = np.random.random((3,3))
    R, _ = np.linalg.qr(A)
    p = np.random.random((3,1))

    print('R:\n', R, '\n')

    Rtoquat = rotm2quat(R)
    print('Rtoquat:\n', Rtoquat, '\n')

    quattoaxang = quat2axang(Rtoquat)
    print('quattoaxang:\n', quattoaxang, '\n')

    axangtorotm = axang2rotm(quattoaxang)
    print('axangtorotm:\n', axangtorotm, '\n')

    rotmtoaxang = rotm2axang(axangtorotm)
    print('rotmtoaxang:\n', rotmtoaxang, '\n')

    axangtoquat = axang2quat(rotmtoaxang)
    print('axangtoquat:\n', axangtoquat, '\n')

    quattorotm = quat2rotm(axangtoquat)
    print('quattorotm:\n', quattorotm, '\n')

    p1 = R@p
    print('p1\n', p1, '\n')

    p2 = quatrotate(axangtoquat, p)
    print('p2\n', p2, '\n')

    # test rvec
    rvec = np.random.random((1,3))
    print('rvec:\n', rvec)

    rvectoaxang = rvec2axang(rvec)
    print('rvectoaxang\n', rvectoaxang)
    axangtorvec = axang2rvec(rvectoaxang)
    print('axangtorvec:\n', axangtorvec)

    rvectoquat = rvec2quat(rvec)
    print('rvectoquat:\n', rvectoquat)
    quattorvec = quat2rvec(rvectoquat)
    print('quattorvec:\n', quattorvec)

    rvectorotm = rvec2rotm(rvec)
    print('rvectortm:\n', rvectorotm)
    rotmtorvec = rotm2rvec(rvectorotm)
    print('rotmtorvec:\n', rotmtorvec)

    rvectorotm_cv, _ = cv2.Rodrigues(rvec)
    print('rvectorotm_cv:\n', rvectorotm_cv)


    time.sleep(1)

    time_begin = time.time()
    for i in range(500000):
        axang = rotm2quat(R)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        axang = quat2axang(Rtoquat)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        R = axang2rotm(quattoaxang)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        axang = rotm2axang(axangtorotm)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        R = axang2quat(rotmtoaxang)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

    time_begin = time.time()
    for i in range(500000):
        R = quat2rotm(axangtoquat)
    time_end = time.time()
    print('average:', (time_end-time_begin)*1000/500000, 'ms')

File: rigidbodytransform/rigidtransform_test.m
%%
clc;
clear;

addpath('fastrigidtransform');


%%
A = rand(3,3)
[R, ~] = qr(A)
p = rand(3,1)

Rtoquat = frotm2quat(R)
quattoaxang = fquat2axang(Rtoquat)
axangtorotm = faxang2rotm(quattoaxang)
rotmtoaxang = frotm2axang(axangtorotm)
axangtoquat = faxang2quat(rotmtoaxang)
quattorotm = fquat2rotm(axangtoquat)

Rtoquat_ = rotm2quat(R)
quattoaxang_ = quat2axang(Rtoquat_)
axangtorotm_ = axang2rotm(quattoaxang_)
rotmtoaxang_ = rotm2axang(axangtorotm_)
axangtoquat_ = axang2quat(rotmtoaxang_)
quattorotm_ = quat2rotm(axangtoquat_)


p1 = R*p
p2 = fquatrotate(axangtoquat, p)


%% test rvec
rvec = rand(1,3)

rvectoaxang = frvec2axang(rvec)
axangtorvec = faxang2rvec(rvectoaxang)

rvectoquat = frvec2quat(rvec)
quattorvec = fquat2rvec(rvectoquat)

rvectorotm = frvec2rotm(rvec)
rotmtorvec = frotm2rvec(rvectorotm)


%%
clc;
clear;
close all;

times = 50000;
axang = rand(1,4);
quat = rand(1,4);
A = rand(3,3);
[R, ~] = qr(A);
rvec = rand(1,3);

time_begin = clock;
for i=1:times
%     x = faxang2quat(quat);
%     x = faxang2rotm(quat);
%     x = faxang2rvec(quat);

%     x = fquat2axang(axang);
%     x = fquat2rotm(axang);

%     x = frotm2axang(R);
%     x = frotm2quat(R);
    x = frotm2rvec(R);
end
time_end = clock;
t1 = etime(time_end,time_begin)*1000/times

time_begin = clock;
for i=1:times
%     x = axang2quat(quat);
%     x = axang2rotm(quat);
%     x = axang2rvec(quat);

%     x = quat2axang(axang);
%     x = quat2rotm(axang);

%     x = rotm2axang(R);
%     x = rotm2quat(R);
end
time_end = clock;
t1 = etime(time_end,time_begin)*1000/times

